xds/src/main/java/io/grpc/xds/XdsDependencyManager.java

```diff
@@ -189,23 +199,17 @@ private void releaseSubscription(ClusterSubscription subscription) {
     checkNotNull(subscription, "subscription");
     String clusterName = subscription.getClusterName();
     syncContext.execute(() -> {
-      Set<ClusterSubscription> subscriptions = clusterSubscriptions.get(clusterName);
-      if (subscriptions == null || !subscriptions.remove(subscription)) {
-        logger.log(DEBUG, "Subscription already released for {0}", clusterName);
-        return;
-      }
-
-      if (subscriptions.isEmpty()) {
-        clusterSubscriptions.remove(clusterName);
-        XdsWatcherBase<?> cdsWatcher =
-            resourceWatchers.get(CLUSTER_RESOURCE).watchers.get(clusterName);
-        cancelClusterWatcherTree((CdsWatcher) cdsWatcher, subscription.toString());
-        maybePublishConfig();
+      XdsWatcherBase<?> cdsWatcher =
+          resourceWatchers.get(CLUSTER_RESOURCE).watchers.get(clusterName);
+      if (cdsWatcher == null) {
+        return; // already released while waiting for the syncContext
```

Comment:
This comment may be out-of-date. Since the subscription hasn't been removed from parent sets, this would only be possible if releaseSubscription() had already been called for this subscription (which is up to you whether is an error or not).